### ISSUE TITLE
dlopen(...../_multiarray_umath.cpython-38-darwin.so, 2): Symbol not found: _npy_cabs

### ISSUE THREAD

**Issue by milosivanovic** [numpy/numpy#18523]
### Environment details:

**OS: Big Sur 11.2.2
Architecture: arm64/M1/Apple Silicon** (no Rosetta2)

Versions:

```
milos@air:~/d$ file $(which python3)
/opt/homebrew/bin/python3: Mach-O 64-bit executable arm64
```

```
milos@air:~/d$ python3 --version
Python 3.8.8
```

```
milos@air:~/d$ which pip3
/opt/homebrew/bin/pip3
```

```
milos@air:~/d$ pip3 list
Package    Version
---------- -------
appdirs    1.4.4
distlib    0.3.1
filelock   3.0.12
pip        21.0.1
setuptools 54.0.0
six        1.15.0
virtualenv 20.4.2
wheel      0.36.2
```

```
milos@air:~/d$ virtualenv --version
virtualenv 20.4.2 from /opt/homebrew/lib/python3.8/site-packages/virtualenv/__init__.py
```

```
milos@air:~/d$ clang -v
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: arm64-apple-darwin20.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

### Reproducing code example:
```
milos@air:~/d$ virtualenv env
created virtual environment CPython3.8.8.final.0-64 in 152ms
  creator CPython3Posix(dest=/Users/milos/d/env, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/Users/milos/Library/Application Support/virtualenv)
    added seed packages: pip==21.0.1, setuptools==53.1.0, wheel==0.36.2
  activators BashActivator,CShellActivator,FishActivator,PowerShellActivator,PythonActivator,XonshActivator

milos@air:~/d$ source env/bin/activate

(env) milos@air:~/d$ pip3 list
Package    Version
---------- -------
pip        21.0.1
setuptools 53.1.0
wheel      0.36.2

(env) milos@air:~/d$ pip3 install --upgrade pip setuptools wheel
Requirement already satisfied: pip in ./env/lib/python3.8/site-packages (21.0.1)
Requirement already satisfied: setuptools in ./env/lib/python3.8/site-packages (53.1.0)
Collecting setuptools
  Using cached setuptools-54.0.0-py3-none-any.whl (784 kB)
Requirement already satisfied: wheel in ./env/lib/python3.8/site-packages (0.36.2)
Installing collected packages: setuptools
  Attempting uninstall: setuptools
    Found existing installation: setuptools 53.1.0
    Uninstalling setuptools-53.1.0:
      Successfully uninstalled setuptools-53.1.0
Successfully installed setuptools-54.0.0

(env) milos@air:~/d$ pip3 list
Package    Version
---------- -------
pip        21.0.1
setuptools 54.0.0
wheel      0.36.2

(env) milos@air:~/d$ git clone https://github.com/numpy/numpy --single-branch --branch v1.20.1
Cloning into 'numpy'...
remote: Enumerating objects: 79, done.
remote: Counting objects: 100% (79/79), done.
remote: Compressing objects: 100% (49/49), done.
remote: Total 169161 (delta 35), reused 34 (delta 30), pack-reused 169082
Receiving objects: 100% (169161/169161), 77.68 MiB | 6.50 MiB/s, done.
Resolving deltas: 100% (134421/134421), done.
Note: switching to 'd7aa4085623b222058edb0ff38392c38c5e00c54'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

(env) milos@air:~/d$ cd numpy

(env) milos@air:~/d/numpy$ git log -n 1
commit d7aa4085623b222058edb0ff38392c38c5e00c54 (HEAD, tag: v1.20.1)
Author: Charles Harris <charlesr.harris@gmail.com>
Date:   Sun Feb 7 11:43:12 2021 -0700

    REL: NumPy 1.20.1 release.

(env) milos@air:~/d/numpy$ pip3 install Cython
Collecting Cython
  Using cached Cython-0.29.22-py2.py3-none-any.whl (980 kB)
Installing collected packages: Cython
Successfully installed Cython-0.29.22

(env) milos@air:~/d/numpy$ pip3 install . --no-binary :all: --no-use-pep517
Processing /Users/milos/d/numpy
Skipping wheel build for numpy, due to binaries being disabled for it.
Installing collected packages: numpy
    Running setup.py install for numpy ... done
Successfully installed numpy-1.20.1

(env) milos@air:~/d/numpy$ cd

(env) milos@air:~$ python3
Python 3.8.8 (default, Feb 26 2021, 09:10:27)
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
Traceback (most recent call last):
  File "/Users/milos/d/env/lib/python3.8/site-packages/numpy/core/__init__.py", line 22, in <module>
    from . import multiarray
  File "/Users/milos/d/env/lib/python3.8/site-packages/numpy/core/multiarray.py", line 12, in <module>
    from . import overrides
  File "/Users/milos/d/env/lib/python3.8/site-packages/numpy/core/overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
ImportError: dlopen(/Users/milos/d/env/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-darwin.so, 2): Symbol not found: _npy_cabs
  Referenced from: /Users/milos/d/env/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-darwin.so
  Expected in: flat namespace
 in /Users/milos/d/env/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-darwin.so

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/milos/d/env/lib/python3.8/site-packages/numpy/__init__.py", line 145, in <module>
    from . import core
  File "/Users/milos/d/env/lib/python3.8/site-packages/numpy/core/__init__.py", line 48, in <module>
    raise ImportError(msg)
ImportError:

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.8 from "/Users/milos/d/env/bin/python3"
  * The NumPy version is: "1.20.1"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: dlopen(/Users/milos/d/env/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-darwin.so, 2): Symbol not found: _npy_cabs
  Referenced from: /Users/milos/d/env/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-darwin.so
  Expected in: flat namespace
 in /Users/milos/d/env/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-darwin.so

```

### Error message:

```
>>> import numpy
Traceback (most recent call last):
  File "/Users/milos/d/env/lib/python3.8/site-packages/numpy/core/__init__.py", line 22, in <module>
    from . import multiarray
  File "/Users/milos/d/env/lib/python3.8/site-packages/numpy/core/multiarray.py", line 12, in <module>
    from . import overrides
  File "/Users/milos/d/env/lib/python3.8/site-packages/numpy/core/overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
ImportError: dlopen(/Users/milos/d/env/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-darwin.so, 2): Symbol not found: _npy_cabs
  Referenced from: /Users/milos/d/env/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-darwin.so
  Expected in: flat namespace
 in /Users/milos/d/env/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-darwin.so
```

### NumPy/Python version information:

Unable to properly import numpy to run `import sys, numpy; print(numpy.__version__, sys.version)` but the version info should be clear from the information I already provided above

### FULL BUILD OUTPUT:
I captured the full build output with `python3 setup.py install > /tmp/full-build-output.txt 2>&1` which is attached below:
[full-build-output.txt](https://github.com/numpy/numpy/files/6065314/full-build-output.txt)

I noticed something interesting in the output, and grepped it here:
```
(env) milos@air:~/d/numpy$ grep 'ld: warning:' /tmp/full-build-output.txt
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpymath.a, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x21 0x3C 0x61 0x72 0x63 0x68 0x3E 0x0A 0x2F 0x20 0x20 0x20 0x20 0x20 0x20 0x20 )
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpymath.a, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x21 0x3C 0x61 0x72 0x63 0x68 0x3E 0x0A 0x2F 0x20 0x20 0x20 0x20 0x20 0x20 0x20 )
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpymath.a, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x21 0x3C 0x61 0x72 0x63 0x68 0x3E 0x0A 0x2F 0x20 0x20 0x20 0x20 0x20 0x20 0x20 )
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpyrandom.a, building for macOS-arm64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpyrandom.a, building for macOS-arm64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpyrandom.a, building for macOS-arm64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpyrandom.a, building for macOS-arm64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpyrandom.a, building for macOS-arm64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpyrandom.a, building for macOS-arm64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpyrandom.a, building for macOS-arm64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpyrandom.a, building for macOS-arm64 but attempting to link with file built for macOS-arm64
```

I wonder if this could be related. The second half of those error messages make no sense ("building for macOS-arm64 but attempting to link with file built for macOS-arm64"? What is that supposed to mean?)

### TL;DR:
I can apparently build numpy "successfully" but I can't import it afterwards since it complains about missing symbols (specifically _npy_cabs). There are some warnings (see above) from the linker, but I can't make sense of them or figure out what to try next. The _npy_cabs symbol can be found when grepping:

```
(env) milos@air:~/d/env/lib/python3.8/site-packages/numpy/core$ grep _npy_cabs _multiarray_umath.cpython-38-darwin.so
grep: _multiarray_umath.cpython-38-darwin.so: binary file matches
(env) milos@air:~/d/env/lib/python3.8/site-packages/numpy/core$ grep _npy_cabs lib/libnpymath.a
grep: lib/libnpymath.a: binary file matches
```

but I'm not sure where they are referenced and where they are missing.

Any help is greatly appreciated!

**Comment by milosivanovic** [comment]
**I figured out the issue thanks to a tip from https://github.com/bitcoin/bitcoin/issues/20825#issuecomment-753444519.**

Prior to wanting to install numpy I had installed many GNU utilities via brew as I prefer to use those over Apple's whenever possible. One of those packages was binutils. The binutils package contains a bunch of GNU binary utils, such as `strings`, `ld`, `as`, `objdump`, etc. In addition to installing the package, I modified my PATH variable to add those utils in front of the system-provided ones: `export PATH="/opt/homebrew/opt/binutils/bin:$PATH"`. This conflicted with Apple's own build tools when the numpy C extensions were being compiled. Specifically, it seems the `ranlib` utility was the culprit. It's a tool included with macOS by default, and it lost presedence after my PATH update:

Before - A system in this state **will not** be able to successfully build and use numpy:
```
(env) milos@air:~/d/numpy$ PATH="/opt/homebrew/opt/binutils/bin:$PATH" which ranlib
/opt/homebrew/opt/binutils/bin/ranlib
```

After - A system in this state **will** be able to successfully build and use numpy:
```
(env) milos@air:~/d/numpy$ which ranlib
/usr/bin/ranlib
```

According to the manpage, `ranlib` is a tool to modify archive libraries:
```
man libtool
LIBTOOL(1)

NAME
       libtool - create libraries
       ranlib - add or update the table of contents of archive libraries
...
```

...which is exactly what `libnpymath.a` is:

```
(env) milos@air:~/d/env/lib/python3.8/site-packages/numpy/core/lib$ file libnpymath.a
libnpymath.a: current ar archive random library
```

and so looking closely at a particular line in the build output:

```
clang -bundle -undefined dynamic_lookup -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk build/temp.macosx-11-arm64-3.8/build/src.macosx-11-arm64-3.8/numpy/core/src/multiarray/_multiarray_tests.o build/temp.macosx-11-arm64-3.8/numpy/core/src/common/mem_overlap.o -L/opt/homebrew/lib -L/opt/homebrew/opt/openssl@1.1/lib -L/opt/homebrew/opt/sqlite/lib -L/opt/homebrew/opt/tcl-tk/lib -Lbuild/temp.macosx-11-arm64-3.8 -lnpymath -o build/lib.macosx-11-arm64-3.8/numpy/core/_multiarray_tests.cpython-38-darwin.so
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpymath.a, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x21 0x3C 0x61 0x72 0x63 0x68 0x3E 0x0A 0x2F 0x20 0x20 0x20 0x20 0x20 0x20 0x20 )
building 'numpy.core._multiarray_umath' extension
compiling C dispatch-able sources
CCompilerOpt.parse_targets[1741] : looking for '@targets' inside ->  build/src.macosx-11-arm64-3.8/numpy/core/src/umath/loops_unary_fp.dispatch.c
CCompilerOpt._parse_target_tokens[1910] : skip targets (SSE2 VSX2) not part of baseline or dispatch-able features
CCompilerOpt._parse_policy_not_keepbase[2022] : skip baseline features (NEON)
CCompilerOpt._parse_target_tokens[1934] : policy 'MAXOPT' is ON
CCompilerOpt._parse_policy_maxopt[2043] : debug mode is detected, policy 'maxopt' is skipped.
CCompilerOpt._generate_config[2469] : generate dispatched config ->  build/src.macosx-11-arm64-3.8/numpy/core/src/umath/loops_unary_fp.dispatch.h
C compiler: clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
```

we see it was specifically trying to link `numpy.core._multiarray_umath` which is the shared object with the missing symbols, and it generated that very suspicious "warning":

```
ld: warning: ignoring file build/temp.macosx-11-arm64-3.8/libnpymath.a, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x21 0x3C 0x61 0x72 0x63 0x68 0x3E 0x0A 0x2F 0x20 0x20 0x20 0x20 0x20 0x20 0x20 )
```

which is actually an extremely important error in my opinion as it basically means the resulting binary will be useless.

So after removing the binutils binaries from my $PATH, the build succeeded without those linker "warnings" and I was able to import numpy:

```
(env) milos@air:~$ python3
Python 3.8.8 (default, Feb 26 2021, 09:10:27)
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys, numpy; print(numpy.__version__, sys.version)
1.20.1 3.8.8 (default, Feb 26 2021, 09:10:27)
[Clang 12.0.0 (clang-1200.0.32.29)]
```

Success!

**Comment by milosivanovic** [comment]
In terms on what we can learn from this experience: would it be possible to convert those linker warnings into errors during the numpy source installation process? Or, some other way to generate a fatal compilation error, seeing as they are critical to a properly-functioning numpy? That way if someone else has a similar incompatible environment we can at least make the problem more visible to them rather than buried in "successful" build logs.

**Comment by VishnuRamD** [comment]
I Couldn't import sklearn. Getting the below error which is similar to yours. Could you help me with this ?
ImportError: dlopen(/opt/anaconda3/lib/python3.8/site-packages/scipy/spatial/qhull.cpython-38-darwin.so, 2): bad rebase type 0 in /opt/anaconda3/lib/python3.8/site-packages/scipy/spatial/../.dylibs/libopenblas.0.dylib

Adding detailed problem statement in below stakeoverflow question
https://stackoverflow.com/questions/69054733/issue-in-qhull-cpython-38-darwin-so-while-importing-sklearn

**Comment by rgommers** [comment]
Thanks for the detailed report @milosivanovic. I agree that this warning should be turned into an error so the build fails. 

Let's try to do this after the change to a new build system.

**Comment by rgommers** [comment]
We've now completed the build system switch, so I revisited this. The problem of `ld: warning: ignoring file` as a result of having a toolchain for the wrong macOS architecture can't be turned into an error it looks like; it's macOS's linker that does this. I found 3 bug reports with this string in the Meson issue tracker, and they are all due to a similar reason, basically having an `arm64` static library and trying to link it into a target for `x86-64` (or vice versa).

I don't think there's anything reasonable we can do here, since the correct static library exists but `ld` decides it's fine to ignore it. This is for the toolchain (macOS or binutils to fix).

I'll go ahead and close this as inactionable. If you think I missed something and see a way to fix this, I'd be happy to reopen and reconsider @milosivanovic.
